task/probe.go

```diff
@@ -0,0 +1,194 @@
+	"strings"
+
+	livepeerAPI "github.com/livepeer/go-api-client"
+	ffprobe "gopkg.in/vansante/go-ffprobe.v2"
+)
+
+var (
+	supportedFormats     = []string{"mp4", "mov"}
+	supportedVideoCodecs = map[string]bool{"h264": true}
+	supportedAudioCodecs = map[string]bool{"aac": true}
+)
+
+type FileMetadata struct {
+	MD5       string                 `json:"md5"`
+	SHA256    string                 `json:"sha256"`
+	Ffprobe   *ffprobe.ProbeData     `json:"ffprobe"`
+	AssetSpec *livepeerAPI.AssetSpec `json:"assetSpec"`
+}
+
+func Probe(ctx context.Context, filename string, data io.Reader) (*FileMetadata, error) {
+	hasher := NewReadHasher(data)
+	probeData, err := ffprobe.ProbeReader(ctx, hasher)
+	if err != nil {
+		return nil, fmt.Errorf("error probing file: %w", err)
+	}
+	if _, err := hasher.FinishReader(); err != nil {
+		return nil, fmt.Errorf("error reading input: %w", err)
+	}
+	md5, sha256 := hasher.MD5(), hasher.SHA256()
+	assetSpec, err := toAssetSpec(filename, probeData, []livepeerAPI.AssetHash{
+		{Hash: md5, Algorithm: "md5"},
+		{Hash: sha256, Algorithm: "sha256"}})
+	if err != nil {
+		return nil, fmt.Errorf("error processing ffprobe output: %w", err)
+	}
+	return &FileMetadata{
+		MD5:       md5,
+		SHA256:    sha256,
+		Ffprobe:   probeData,
+		AssetSpec: assetSpec,
+	}, nil
+}
+
+func toAssetSpec(filename string, probeData *ffprobe.ProbeData, hash []livepeerAPI.AssetHash) (*livepeerAPI.AssetSpec, error) {
+	if filename == "" && probeData.Format.Filename != "pipe:" {
+		filename = probeData.Format.Filename
+	}
+	format, err := findFormat(supportedFormats, probeData.Format.FormatName, filename)
+	if err != nil {
+		return nil, err
+	}
+	bitrate, err := strconv.ParseFloat(probeData.Format.BitRate, 64)
+	if probeData.Format.BitRate != "" && err != nil {
+		return nil, fmt.Errorf("error parsing file bitrate: %w", err)
+	}
+	size, err := strconv.ParseInt(probeData.Format.Size, 10, 64)
+	if probeData.Format.Size != "" && err != nil {
+		return nil, fmt.Errorf("error parsing file size: %w", err)
+	}
+	spec := &livepeerAPI.AssetSpec{
+		Name: filename,
+		Hash: hash,
+		Size: size,
+		VideoSpec: &livepeerAPI.AssetVideoSpec{
+			Format:      format,
+			DurationSec: probeData.Format.DurationSeconds,
+			Bitrate:     bitrate,
+			Tracks:      make([]*livepeerAPI.AssetTrack, 0, len(probeData.Streams)),
+		},
+	}
+	if ns := len(probeData.Streams); ns > 2 {
+		return nil, fmt.Errorf("too many tracks in file: %d", ns)
+	}
+	var hasVideo bool
+	for _, stream := range probeData.Streams {
+		track, err := toAssetTrack(stream)
+		if err != nil {
+			return nil, err
+		}
+		hasVideo = hasVideo || track.Type == "video"
+		spec.VideoSpec.Tracks = append(spec.VideoSpec.Tracks, track)
+	}
+	if !hasVideo {
+		return nil, fmt.Errorf("no video track found in file")
+	}
+	return spec, nil
+}
+
+func findFormat(supportedFormats []string, format, filename string) (string, error) {
+	actualFormats := strings.Split(format, ",")
+	extension := path.Ext(filename)
```

Comment:
The file extension comes either from the URL or the `Content-Disposition` header from the response. It has no commitment with being truthful, as I could just rename any text file to be mp4 if I want. The only really trustable format  is the one that comes out of the probe.

The problem is that the output from the probe is not very specific. The format both from `mp4` or `mov` files come out as just `"format_name": "mov,mp4,m4a,3gp,3g2,mj2"` from `ffprobe`. That's why I'm using the extension here, to pick exactly the `extension` format, if it is accurate to what the file actually is. Whole logic is:
 - spit that list of formats from the probe into all the individual formats
 - check if any of those formats are supported + match the file extension. If they do, that's what we're looking for and we return the extension.
 - If the extension doesn't match any actual file format that is suppoted, fallback to the first supported format that shows up in the "actual formats" of the file. In that case, we'll always fallback to `mp4` since it's our top preferred format (first element in the supported formats array)